include/property.h

```diff
@@ -0,0 +1,181 @@
+#ifndef __PROPERTY_H__
+#define __PROPERTY_H__
+
+#include <iostream>
+#include <set>
+#include <tuple>
+#include <sstream>
+#include <regex>
+
+namespace clayer {
+
+    // Concepts used for property reading and writing
+    template<typename T>
+    concept bool Streamable = requires(T o, std::ostream& s) {
+        {s << o} -> std::ostream&;
+    };
+
+    template<typename T>
+    concept bool StringRef = requires(T o) {
+        {*o} -> std::string;
+    };
+
+    // template<typename T=void>
+    // void print<>(){
+    //     std::cout << "\n";
+    // }
+
+    // template<Streamable T, typename... Targs>
+    // void print(T o, Targs... tail) {
+    //     std::cout << o <<" ";
+    //     print(tail...);
+    // }
+
+    // ## Log LogRecord - Any item that can be printed in a Log Record
+    enum log_properties {
+        FILE,   // - File name (CodeContext)
+        FUNC,   // - Function name (CodeContext)
+        LINE,   // - Line number (CodeContext)
+        // LEVEL
+        // NAME
+        // THREAD
+        // HASH
+        DATE,   // - Date a Log Record is created (RunContext)
+        TIME,   // - Time a Log Record is created (RunContext)
+        MESG    // - Added by the user
+    };
+
+    // ### Code Context - Log Properties that can be inferred from Code
+    struct CodeContext {
+        std::string file;
+        std::string func;
```

Comment:
CodeContext is intended for both analyser and logger both. However I have just implemented the bare necessary for analyser. Given that the logger is working in it's current state, We may not want to change it.
also, between analyser and logger, the major difference is that logger needs constant strings (or pointeres to const strings, whatever -- not too worried about that, according Bjarne's principles, strings may not be too far from pointers to characters). That difference of constant and non constant strings can be captured by making CodeContext a const.
So, from my perspective, the next steps here are:
1. Replace `CodeInfo` in analyser with `const CodeContext`
2. Make any changes that might be necessary to get this working on both logger and analyser.
3. Measure the difference in performance for logger for CodeInfo (with Pointers to char array) vs const CodeContext (with strings) in O2/O3.
4. If the difference is big, implement const char in CodeContext, or live with CodeInfo.

Given that the whole excercise could lead to the same state as now, we may want to tackle this in v1.2